shdlc/sensirion_shdlc.c

```diff
@@ -201,3 +201,184 @@ int16_t sensirion_shdlc_rx(uint8_t max_data_len,
+                                         (uint8_t)((data & 0x00FF0000) >> 16));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x0000FF00) >> 8));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x000000FF) >> 0));
+}
+
+void sensirion_shdlc_add_int32_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, int32_t data) {
+
+    sensirion_shdlc_add_uint32_t_to_frame(tx_frame, (uint32_t)data);
+}
+
+void sensirion_shdlc_add_uint16_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, uint16_t data) {
+
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0xFF00) >> 8));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x00FF) >> 0));
+}
+
+void sensirion_shdlc_add_int16_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, int16_t data) {
+
+    sensirion_shdlc_add_uint16_t_to_frame(tx_frame, (uint16_t)data);
+}
+
+void sensirion_shdlc_add_float_to_frame(struct sensirion_shdlc_buffer* tx_frame,
+                                        float data) {
+    union {
+        uint32_t uint32_data;
+        float float_data;
+    } convert;
+
+    convert.float_data = data;
+    sensirion_shdlc_add_uint32_t_to_frame(tx_frame, convert.uint32_data);
+}
+
+void sensirion_shdlc_add_bytes_to_frame(struct sensirion_shdlc_buffer* tx_frame,
+                                        uint8_t* data, uint16_t data_length) {
+    uint16_t i;
+
+    for (i = 0; i < data_length; i++) {
+        sensirion_shdlc_add_uint8_t_to_frame(tx_frame, data[i]);
+    }
+}
+
+int16_t sensirion_shdlc_tx_frame(struct sensirion_shdlc_buffer* tx_frame) {
+
+    int16_t tx_length;
+
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame, ~(tx_frame->checksum));
+    tx_frame->data[tx_frame->offset++] = SHDLC_STOP;
+    tx_length = sensirion_uart_hal_tx(tx_frame->offset, tx_frame->data);
+    if (tx_length < 0) {
+        return tx_length;
+    }
+    if (tx_length != tx_frame->offset) {
+        return SENSIRION_SHDLC_ERR_TX_INCOMPLETE;
+    }
+    return NO_ERROR;
+}
+
+static uint8_t
+sensirion_shdlc_unstuff_next_byte(struct sensirion_shdlc_buffer* rx_frame) {
+
+    uint8_t data = rx_frame->data[rx_frame->offset++];
+
+    if (data == 0x7d) {
+        data = rx_frame->data[rx_frame->offset++];
+        switch (data) {
+            case 0x31:
+                rx_frame->checksum += 0x11;
+                return 0x11;
+            case 0x33:
+                rx_frame->checksum += 0x13;
+                return 0x13;
+            case 0x5d:
+                rx_frame->checksum += 0x7d;
+                return 0x7d;
+            case 0x5e:
+                rx_frame->checksum += 0x7e;
+                return 0x7e;
+            default:
+                rx_frame->checksum += data;
+                return data;
```

Comment:
Instead of the switch case I'd just flip the one bit that is required. Also the case that a byte gets stuffed which shouldn't would be an error case IMO.
See also https://github.com/Sensirion/embedded-common/blob/37e0dab65f2db700a425c7d91ca307a42d3a956b/shdlc/sensirion_shdlc.c#L72

Reply:
Yes that would be better code and if it goes wrong it's and error.

Reply:
See: https://github.com/Sensirion/embedded-common/pull/65/commits/270a67fef52524322cf15ec6685188984e9b8477